### PR TITLE
Add fact parser validation, tests, and bulk fact loading with tests

### PR DESCRIPTION
**Pull request by ColtonPayne**
## Summary

This PR adds comprehensive input validation to the fact parser and implements bulk fact loading from CSV files with extensive test coverage.

### Fact Parser Validation (`fact_parser.py`) (Issue #91 )
- Added validation for empty/whitespace-only input 
- Validates parentheses structure and placement (Issue #89 )
- Enforces valid predicate naming (must start with letter or underscore, alphanumeric + underscore allowed)
- Validates component structure (no nested parentheses, colons, etc.)
- Validates interval bounds are within [0, 1] range and are in the proper format (issue #90 )
- Validates interval lower <= upper bound
- Prevents double negation and negation with explicit bounds
- Provides clear, specific error messages for each validation failure

### Bulk Fact Loading (`pyreason.py`)
- Implemented `add_fact_in_bulk()` function to load facts from CSV files
- Supports optional header row detection
- Handles optional columns: name, start_time, end_time, static
- Provides warnings for invalid data (malformed facts, invalid times, etc.) without crashing
- Supports multiple boolean formats for static field (True/true/1/yes, False/false/0/no)

### Test Coverage
- **333 new lines** in `test_fact_parser.py`:
  - Tests for valid fact parsing (node/edge facts, intervals, negation, etc.)
  - Tests for invalid inputs (missing parentheses, empty fields, invalid characters, etc.)
  - Edge cases and boundary conditions
- **204 new lines** in `test_pyreason_file_loading.py`:
  - Tests for bulk fact loading from CSV
  - Warning validation for invalid facts
  - Tests with/without headers, various static value formats
  - Error handling tests

### Test Data
- Created `example_facts.csv` with comprehensive test scenarios including both valid and invalid facts
- Created `example_facts_no_header.csv` for testing CSV without headers

### Implementation Notes
- All validation errors raise `ValueError` with descriptive messages
- Bulk loading continues processing even when individual rows fail (with warnings)
- Predicate validation regex: `^[a-zA-Z_][a-zA-Z0-9_]*$` (follows Python identifier rules)

🤖 Generated with Claude Code

### CODE REVIEW

**Review by dyumanaditya — DISMISSED** [review]
@ColtonPayne Everything looks good except for my one comment on explicit bound inverses.

File: pyreason/pyreason.py [hunk 1]
```diff
@@ -648,125 +649,157 @@ def add_fact(pyreason_fact: Fact) -> None:
         __edge_facts.append(f)
 
 
-def add_fact_in_bulk(csv_path: str, raise_errors = True) -> None:
-    """Load multiple facts from a CSV file.
+def add_fact_in_bulk(json_path: str, raise_errors = True) -> None:
+    """Load multiple facts from a JSON file.
 
-    The CSV should have columns representing Fact attributes in this order:
```
Comment (kmukherji) [inline pyreason/pyreason.py]: I think keep load_facts_from_csv as a method. JSON should be default, but since we have built the functionality for CSV already, keep that in. Someone may already have a working project that uses CSV input. Someone may download a dataset in CSV format etc.

File: pyreason/pyreason.py [hunk 2]
```diff
@@ -648,125 +649,157 @@ def add_fact(pyreason_fact: Fact) -> None:
...
+                error_count += 1
+                continue
+
+            fact_text = str(fact_text).strip()
 
             # Extract optional parameters with defaults
-            name = row[1].strip() if len(row) > 1 and row[1].strip() else None
+            name = fact_obj.get('name')
+            if name is not None:
+                name = str(name).strip() if str(name).strip() else None
```
Comment (kmukherji) [inline pyreason/pyreason.py]: How do we want to handle duplicates here?

**Review by kmukherji — CHANGES_REQUESTED** [review]
Some more comments are inline.

File: pyreason/pyreason.py (hunk 2 above)
Comment (ColtonPayne) [reply pyreason/pyreason.py]: This is a good question - I added a check that requires all fact names in the json/csv to be distinct.  
  There is no such validation in the regular add_fact() function, so there is nothing stopping a user 
  from adding multiple facts with the same name using the traditional method of loading facts.  We    
  could add a check in pyreason.py to prohibit this as well, but it would be a breaking change for    
  anyone who is currently running experiments with facts that have duplicate names.  Because fact     
  names are only used in the atom trace and not in the fp operator for convergence, having duplicate  
  names will not impact correctness and would just leave to a confusing atom trace.  I decided to handle this discrepancy by adding a warning to the add_fact method for users who add duplicate facts this way.

File: pyreason/scripts/facts/fact.py [hunk 3]
```diff
@@ -6,7 +6,45 @@ class Fact:
...
+            - `'~pred(node):[0.2,0.8]'` - negation with explicit bound
+            NOTE: Negating an explicit bound will round the upper and lower bounds to 10 decimal places before taking the negation
+            This is needed to avoid floating point precision errors.
+
+            **Invalid examples:**
+            - `'123pred(node)'` - predicate starts with digit
+            - `'pred@name(node)'` - invalid characters in predicate
+            - `'pred(node1,node2,node3)'` - more than 2 components
+            - `'pred(node):[1.5,2.0]'` - values out of range [0,1]
+
```
Comment (kmukherji) [inline pyreason/scripts/facts/fact.py]: We should follow the conventions of the logic community across all our examples.
All Predicates and Variables start with capital letters. All constants (even proper nouns like people's names) should be in small letters (e.g. alice, bob).

In logic, predicates cannot start with underscores, however since we are not protecting it for any other use - for now we can keep it as is. Only update the example here.

File: pyreason/scripts/utils/fact_parser.py [hunk 4]
```diff
@@ -148,6 +153,11 @@ def parse_fact(fact_text):
         if lower > upper:
             raise ValueError(f"Interval lower bound {lower} cannot be greater than upper bound {upper}")
 
-        bound = interval.closed(*bound_values)
+        # We calculate ~[l,u] = [1-u, 1-l]
+        # Round to eliminate floating point precision errors (e.g., 1 - 0.8 = 0.19999999...)
+        if negate_interval:
+            lower, upper = round(1 - upper, 10), round(1 - lower, 10)
+
+        bound = interval.closed(lower, upper)
```
[the remark this answers is not in the document]
Comment (kmukherji) [reply pyreason/scripts/utils/fact_parser.py]: I think we should definitely support at least two decimal points. Maybe 3 - unless that detoriorates performance.

File: pyreason/pyreason.py [hunk 5]
```diff
@@ -639,15 +705,272 @@ def add_fact(pyreason_fact: Fact) -> None:
     if pyreason_fact.type == 'node':
         if pyreason_fact.name is None:
             pyreason_fact.name = f'fact_{len(__node_facts)+len(__edge_facts)}'
+
+        if pyreason_fact.name in __node_facts_name_set:
+            warnings.warn(f"Fact {pyreason_fact.name} has already been added. Duplicate fact names will lead to an ambiguous node and atom traces.")
+
```
Comment (kmukherji) [inline pyreason/pyreason.py]: See my other comment about renaming these while loading.

File: pyreason/pyreason.py [hunk 6]
```diff
@@ -639,15 +705,272 @@ def add_fact(pyreason_fact: Fact) -> None:
...
         f = fact_edge.Fact(pyreason_fact.name, pyreason_fact.component, pyreason_fact.pred, pyreason_fact.bound, pyreason_fact.start_time, pyreason_fact.end_time, pyreason_fact.static)
+        __node_facts_name_set.add(pyreason_fact.name)
         __edge_facts.append(f)
 
 
+def add_fact_from_json(json_path: str, raise_errors = True) -> None:
+    """Load multiple facts from a JSON file.
+
+    The JSON should be an array of objects, where each object represents a Fact with these fields:
+    - fact_text (required): The fact in text format, e.g., 'pred(x,y) : [0.2, 1]' or 'pred(x) : True'
```
Comment (kmukherji) [inline pyreason/pyreason.py]: See other comment about conventions.

**Review by kmukherji — COMMENTED** [review]
Once these comments are addressed we can merge this branch.

File: pyreason/pyreason.py [hunk 7]
```diff
@@ -623,6 +626,69 @@ def add_rules_from_file(file_path: str, infer_edges: bool = False) -> None:
...
+        start_time = 0
+
+    # Parse end_time
+    try:
+        end_time = int(end_time_raw) if end_time_raw is not None and str(end_time_raw).strip() else 0
+    except (ValueError, TypeError, AttributeError):
+        if raise_errors:
+            raise ValueError(f"{item_label} {idx}: Invalid end_time '{end_time_raw}'")
+        warnings.warn(f"{item_label} {idx}: Invalid end_time '{end_time_raw}', using default value")
+        end_time = 0
```
[the remark this answers is not in the document]
Comment (ColtonPayne) [reply pyreason/pyreason.py]: We are currently handling invalid values for start_time and end_time to zero if the user passes a non-integer value.  This is the default value provided by the **Fact** constructor.
```
 # Parse start_time
    try:
        start_time = int(start_time_raw) if start_time_raw is not None and str(start_time_raw).strip() else 0
```
If `raise_errors` is false, we handle this gracefully.  If `raise_errors` is true, we will not accept an invalid json/csv and force the user to correct the bad input.

File: pyreason/pyreason.py (hunk 5 above)
Comment (ColtonPayne) [reply pyreason/pyreason.py]: Modifying existing fact loading behavior, even for something trivial like this, would require a major version update.  If we are going to do a major version update (which we should at some point), raising an error is better than modifying the user's facts for them because it will require them to understand the issue before they run their code.  Until we want to do this, a warning is the best we can do as we cannot disrupt existing processes without a major version upgrade.

**Review by kmukherji — APPROVED** [review]
